### PR TITLE
chore: code size improvements

### PR DESCRIPTION
**Pull request by jackmu95**
### Describe the PR

This PR makes some adjustments to the code to decrease overall bundle size (specifically after minimization).

### PR checklist

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix (fixes a boo-boo in the code) - `fix(...)`, requires a patch version update
- [ ] Feature (adds a new feature to BootstrapVue) - `feat(...)`, requires a minor version update
- [x] Enhancement (augments an existing feature) - `feat(...)`, requires a minor version update
- [ ] ARIA accessibility (fixes or improves ARIA accessibility) - `fix(...)`, requires a patch or minor version update
- [ ] Documentation update (improves documentation or typo fixes) - `chore(docs)`, requires a patch version update
- [x] Other (please describe) refactor

**Does this PR introduce a breaking change?** (check one)

- [x] No
- [ ] Yes (please describe since breaking changes require a minor version update)

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, **not** the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (i.e. `[...] (fixes #xxx[,#xxx])`, where "xxx" is the issue number)
- [x] It should address only one issue or feature. If adding multiple features or fixing a bug and adding a new feature, break them into separate PRs if at all possible.
- [x] The title should follow the [**Conventional Commits**](https://www.conventionalcommits.org/) naming convention (i.e. `fix(alert): not alerting during SSR render`, `docs(badge): update pill examples`, `chore(docs): fix typo in README`, etc). **This is very important, as the `CHANGELOG` is generated from these messages, and determines the next version type (patch or minor).**

**If new features/enhancement/fixes are added or changed:**

- [ ] Includes documentation updates
- [ ] Includes component package.json meta section updates (prop, slot and event changes/updates)
- [ ] Includes any needed TypeScript declaration file updates
- [ ] New/updated tests are included and passing (required for new features and enhancements)
- [x] Existing test suites are passing
- [x] CodeCov for patch has met target (all changes/updates have been tested)
- [x] The changes have not impacted the functionality of other components or directives
- [ ] ARIA Accessibility has been taken into consideration (Does it affect screen reader users or keyboard only users? Clickable items should be in the tab index, etc.)

**If adding a new feature, or changing the functionality of an existing feature, the PR's
description above includes:**

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: src/utils/number.spec.js [hunk 1]
```diff
@@ -69,4 +69,18 @@ describe('utils/number', () => {
     expect(toFixed('1 foo', null)).toBe('1')
     expect(toFixed('foo 1 bar', {})).toBe('NaN')
   })
+
+  it('isNumeric() works', async () => {
+    expect(isNumeric(1)).toBe(true)
+    expect(isNumeric('1')).toBe(true)
```
Comment (tmorehouse) [inline src/utils/number.spec.js:75]: Should we test for `0` as well maybe?

```js
expect(isNumeric(0)).toBe(true)
expect(isNumeric('0')).toBe(true)
```

File: src/components/alert/alert.js [hunk 2]
```diff
@@ -150,18 +152,17 @@ export const BAlert = /*#__PURE__*/ Vue.extend({
...
         'div',
         {
-          key: this._uid,
-          staticClass: 'alert',
+          staticClass: CLASS_NAME,
           class: {
-            'alert-dismissible': this.dismissible,
-            [`alert-${this.variant}`]: this.variant
+            [`${CLASS_NAME}-dismissible`]: dismissible,
+            [`${CLASS_NAME}-${variant}`]: !!variant
```
Comment (tmorehouse) [inline src/components/alert/alert.js:158]: Could we maybe make a utility to generate the concatenated class name?

```js
const suffixClass = (base, suffix) => {
  return [base, suffix].filter(identity).join('-')
}
```

Which would reduce the number of transpiled concats: `'base'.concat('-dismissable')`?

Comment (tmorehouse) [reply src/components/alert/alert.js:158]: or maybe:

```js
const suffixClass = (...strings) => {
  return strings.filter(identity).join('-')
}
```

Comment (jackmu95) [reply src/components/alert/alert.js:158]: Place it inside dom utils?

Comment (tmorehouse) [reply src/components/alert/alert.js:158]: Hmmm.... or maybe a generic string utility in `utils/string`?

Comment (tmorehouse) [reply src/components/alert/alert.js:158]: maybe dom, and call it `buildClass`?

Comment (tmorehouse) [reply src/components/alert/alert.js:158]: or `classConcat`

**Comment by tmorehouse** [comment]
Side thought:  Bundlewatch appears to always compare to the base branch `master`

```
[ERROR] Unable to fetch fileDetails for baseBranch=master from https://service.bundlewatch.io/store code=Request failed with status code 404
```

Is it possible to get it to compare aginst the target branch of the PR?

i.e. this PR would compare against `dev`, while the release PR would compare against `master`?

File: src/utils/string.spec.js [hunk 3]
```diff
@@ -75,4 +83,13 @@ describe('utils/string', () => {
   "bar"
 ]`)
   })
+
+  it('suffixClass() works', async () => {
+    const base = 'foo'
+    expect(suffixClass(base, 'bar')).toBe('foo-bar')
+    expect(suffixClass(base, '')).toBe('foo-')
```
Comment (tmorehouse) [inline src/utils/string.spec.js:90]: This would should filter out the empty suffix and be just `foo`?

File: src/utils/string.spec.js [hunk 4]
```diff
@@ -75,4 +83,13 @@ describe('utils/string', () => {
...
+    expect(suffixClass(base, null)).toBe('foo-')
+    expect(suffixClass(base, 0)).toBe('foo-0')
```
Comment (tmorehouse) [inline src/utils/string.spec.js:92]: Same here?

File: src/utils/string.js [hunk 5]
```diff
@@ -62,3 +62,6 @@ export const lowerCase = str => toString(str).toLowerCase()
 
 // Upper case a string
 export const upperCase = str => toString(str).toUpperCase()
+
+// Suffix a class name
+export const suffixClass = (base, suffix) => [base, toString(suffix)].join('-')
```
Comment (tmorehouse) [inline src/utils/string.js:67]: `[base, toString(suffix)].filter(identity).join('-')` ?

File: src/utils/string.spec.js (hunk 3 above)
Comment (jackmu95) [reply src/utils/string.spec.js:90]: I wanted this behavior to prevent that we get only the base branch back.
But nor really sure on that...

Comment (tmorehouse) [reply src/utils/string.spec.js:90]: It is probably never going to happen, as we always text for truthyness of the suffix

Comment (tmorehouse) [reply src/utils/string.spec.js:90]: So maybe it is OK as is.

Comment (tmorehouse) [reply src/utils/string.spec.js:90]: Lets leave it as is without filtering....

Comment (jackmu95) [reply src/utils/string.spec.js:90]: Yeah. I think it's fine like that for now.

**Comment by jackmu95** [comment]
> Side thought: Bundlewatch appears to always compare to the base branch `master`
> 
> ```
> [ERROR] Unable to fetch fileDetails for baseBranch=master from https://service.bundlewatch.io/store code=Request failed with status code 404
> ```
> 
> Is it possible to get it to compare aginst the target branch of the PR?
> 
> i.e. this PR would compare against `dev`, while the release PR would compare against `master`?
> 
> Or maybe generate two bundle watches... one that compares with master and one that compares with dev?

Fixed that by setting the `CI_BRANCH_BASE` env variable to the `github.base_ref` since BundleWatch can't auto-detect that for GitHub Actions.

File: src/components/breadcrumb/breadcrumb-link.js [hunk 6]
```diff
@@ -36,7 +37,7 @@ export const BBreadcrumbLink = /*#__PURE__*/ Vue.extend({
       componentData.attrs = { 'aria-current': suppliedProps.ariaCurrent }
     }
 
-    if (!children) {
+    if (!children || (isArray(children) && children.length === 0)) {
```
Comment (tmorehouse) [inline src/components/breadcrumb/breadcrumb-link.js:40]:
```js
const hasNoKids = children => !children || (isArray(children) && children.length === 0)
```

File: src/components/calendar/calendar.js [hunk 7]
```diff
@@ -1027,14 +1086,21 @@ export const BCalendar = Vue.extend({
       // Return the week "row"
       // We use the first day of the weeks YMD value as a
       // key for efficient DOM patching / element re-use
-      return h('div', { key: week[0].ymd, staticClass: 'row no-gutters' }, $cells)
+      return h(
+        'div',
+        {
+          key: week[0].ymd,
+          staticClass: [CLASS_NAME_ROW, CLASS_NAME_NO_GUTTERS]
```
Comment (tmorehouse) [inline src/components/calendar/calendar.js:1134]: `staticClass` should be a string.

```js
staticClass: [CLASS_NAME_ROW, CLASS_NAME_NO_GUTTERS].join(' ')
```

File: src/components/carousel/carousel.js [hunk 8]
```diff
@@ -564,29 +581,29 @@ export const BCarousel = /*#__PURE__*/ Vue.extend({
...
           'aria-label': this.labelIndicators,
-          'aria-owns': this.safeId('__BV_inner_')
+          'aria-owns': this.safeId('_BV_inner_')
         }
       },
       this.slides.map((slide, n) => {
         return h('li', {
           key: `slide_${n}`,
-          class: { active: n === this.index },
+          class: { [CLASS_NAME_ACTIVE]: n === this.index },
```
Comment (tmorehouse) [inline src/components/carousel/carousel.js:598]: Perhaps switch from Dynamic property names (when the property name is not a computation) to array format:

```js
class: [n === this.index ? CLASS_NAME_ACTIVE : null]
```

As dynamic object properties transpile out to a bit longer code (except if browserlistrc excludes older browsers that do not support dynamic property names)

**Comment by tmorehouse** [comment]
Looks like we need to tweak a few things due to merge conflicts with `dev`

**Comment by jackmu95** [comment]
> @jackmu95 Looks like we need to tweak a few things due to merge conflicts with `dev`

Done!

**Pull request closed without merging**